### PR TITLE
Implement generic DFD decoder - Refers #119

### PR DESCRIPTION
**Pull request by MaciejBaj**
Refers #119

### CODE REVIEW

File: circuit/primitives/src/event_signature.rs [hunk 1]
```diff
@@ -170,4 +237,89 @@ pub mod tests {
...
+        let valid_dfd = "(A(B(C)))";
+        // Important! If using .encode() instead of .as_bytes() + .to_vec(),
+        //  SCALE adds additional byte "92" to event name
+        let decode_res = decode_dfd(valid_dfd.as_bytes().to_vec());
+        assert_eq!(
+            decode_res,
+            Ok(vec![
+                vec![b"A".to_vec()],
+                vec![b"B".to_vec()],
+                vec![b"C".to_vec()]
```
Comment (zannis) [inline circuit/primitives/src/event_signature.rs:305]: Shouldn't this be in the reverse order?

File: circuit/primitives/src/event_signature.rs [hunk 2]
```diff
@@ -170,4 +237,89 @@ pub mod tests {
...
+        let valid_dfd = "(D(C(A,B)))";
+        // Important! If using .encode() instead of .as_bytes() + .to_vec(),
+        //  SCALE adds additional byte "92" to event name
+        let decode_res = decode_dfd(valid_dfd.as_bytes().to_vec());
+        assert_eq!(
+            decode_res,
+            Ok(vec![
+                vec![b"D".to_vec()],
+                vec![b"C".to_vec()],
+                vec![b"A".to_vec(), b"B".to_vec()]
```
Comment (zannis) [inline circuit/primitives/src/event_signature.rs:321]: Same here, should be reverse

**Review by zannis — CHANGES_REQUESTED** [review]

File: circuit/primitives/src/event_signature.rs (hunk 1 above)
Comment (MaciejBaj) [reply circuit/primitives/src/event_signature.rs:305]: Yep!

File: circuit/primitives/src/event_signature.rs (hunk 2 above)
Comment (MaciejBaj) [reply circuit/primitives/src/event_signature.rs:321]: For parallel events won't make a difference

File: circuit/primitives/src/event_signature.rs [hunk 3]
```diff
@@ -170,4 +237,89 @@ pub mod tests {
...
+        let decode_res = decode_dfd(valid_dfd.as_bytes().to_vec());
+        assert_eq!(
+            decode_res,
+            Err("DFD Decoder - attempt to edit step at incorrect depth")
+        )
+    }
+
+    #[test]
+    fn successfully_decodes_dfd_for_3_sequential_events() {
+        let valid_dfd = "(A(B(C)))";
```
Comment (vedhavyas) [inline circuit/primitives/src/event_signature.rs:296]: can we test with some actual examples? Like transfer, swap etc... ?

Comment (MaciejBaj) [reply circuit/primitives/src/event_signature.rs]: added a test case for 32b long values that are IDs

**Pull request merged**